cmd/adapter/adapter.go

```diff
@@ -0,0 +1,430 @@
+	if err == nil {
+		grLogEntries, err := ad.toGrLogEntries(rs.Events)
+		if err == nil {
+			bb, err := json.Marshal(grLogEntries)
+			if err == nil {
+				_, err = rw.Write(bb)
+			}
+		}
+	}
+
+	if err != nil {
+		ad.logger.Error("log(): Err=", err)
+		rw.WriteHeader(http.StatusInternalServerError)
+		_, _ = rw.Write([]byte("oops, something went wrong!\n"))
+	}
+
+	return err
+}
+
+func (ad *Adapter) toGrLogEntries(evs []*api.LogEvent) ([]string, error) {
+	gle := make([]string, 0, len(evs))
+	le := &grLogEntry{Type: "data"}
+	for _, e := range evs {
+		le.Payload = e.Message
+		bb, err := json.Marshal(le)
+		if err != nil {
+			return nil, err
+		}
+		gle = append(gle, string(bb))
+	}
+	return gle, nil
+}
+
+func (ad *Adapter) downloadHandler(ctx context.Context,
+	rw http.ResponseWriter, rq *http.Request, p httprouter.Params) error {
+
+	gzWriter := gzip.NewWriter(rw)
+	defer gzWriter.Close()
+	tarWriter := tar.NewWriter(gzWriter)
+	defer tarWriter.Close()
+
+	lql := buildLql("", ad.cfg.Logrange.Partition, dlLimitLines, -dlLimitLines)
+	ad.logger.Info("download(): Query=", lql)
+
+	var (
+		fnCntr = 0
+		err    error
+	)
+
+	buf := bytes.Buffer{}
+	err = api.Select(ctx, ad.lCli, &api.QueryRequest{Query: lql,
+		Limit: dlLimitLines}, false, func(res *api.QueryResult) {
+		if err == nil {
+			marshal(res.Events, &buf)
+			if buf.Len() > dlLimitPerFileBt {
+				err = writeTar(fnCntr, buf.Bytes(), tarWriter)
+				buf.Reset()
+				fnCntr++
+			}
+		}
+	})
+
+	if err == nil {
+		if buf.Len() > 0 {
+			err = writeTar(fnCntr, buf.Bytes(), tarWriter)
+		}
+	}
+
+	if err != nil {
+		ad.logger.Error("download(): Err=", err)
+		_ = writeTar(fnCntr, []byte("oops, something went wrong!\n"), tarWriter)
+	}
+
+	rw.Header().Set("Content-Disposition", "attachment; filename=logs.tar.gz")
+	return nil
+}
+
+func (ad *Adapter) makeHandlerWithCtx(ctx context.Context, handler handlerWithCtx) httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+		_ = handler(ctx, w, r, p)
+	}
+}
+
+//===================== adapter.cron =====================
+
+func (ad *Adapter) runCronQueries(ctx context.Context) {
+	if ad.cfg == nil || ad.cfg.Logrange == nil {
+		ad.logger.Warn("No config or Logrange config, skipping running cron queries")
+		return
+	}
+
+	ad.logger.Info("Running ", len(ad.cfg.Logrange.CronQueries), " cron queries")
+	for _, cq := range ad.cfg.Logrange.CronQueries {
+		ad.waitWg.Add(1)
+		go func(cq cronQuery) {
+			defer ad.waitWg.Done()
+			ad.runCronQuery(ctx, cq)
+		}(cq)
+	}
+}
+
+func (ad *Adapter) runCronQuery(ctx context.Context, cq cronQuery) {
+	ad.logger.Info("Entering into the loop to run \"", cq.Query, "\" every ", cq.PeriodSec, " seconds.")
+	for {
+		sleepDelay := time.Second * time.Duration(cq.PeriodSec)
+		res, err := ad.lCli.Execute(ctx, api.ExecRequest{Query: cq.Query})
+		if err != nil {
+			ad.logger.Warn("Could not connect to the server to run ", cq)
+			sleepDelay = time.Second
```

Comment:
I would have expected some sort of backoff instead of acceleration in case of errors to be honest.